File: lib/model/services/remote/office_hour_service.dart
import 'package:dio/dio.dart';
import 'package:pbl6mobile/model/entities/office_hour.dart';
import 'package:pbl6mobile/model/services/remote/auth_service.dart';

class OfficeHourService {
  final Dio _dio = AuthService.getSecureDioInstance();

  Future<List<OfficeHour>?> getOfficeHours({
    String? doctorId,
    String? workLocationId,
  }) async {
    final queryParams = <String, dynamic>{};
    if (doctorId != null) queryParams['doctorId'] = doctorId;
    if (workLocationId != null) queryParams['workLocationId'] = workLocationId;

    final response = await _dio.get(
      '/office-hours',
      queryParameters: queryParams,
    );

    if (response.statusCode == 200) {
      final data = response.data;
      if (data['success'] == true) {
        final groupedData = data['data'];
        // Flatten the grouped response similar to React hook
        final List<dynamic> allOfficeHours = [
          ...groupedData['global'] ?? [],
          ...groupedData['workLocation'] ?? [],
          ...groupedData['doctor'] ?? [],
          ...groupedData['doctorInLocation'] ?? [],
        ];

        return allOfficeHours.map((json) => OfficeHour.fromJson(json)).toList();
      }
    }
    return null;
  }

  Future<OfficeHour?> createOfficeHour({
    String? doctorId,
    String? workLocationId,
    required int dayOfWeek,
    required String startTime,
    required String endTime,
    bool isGlobal = false,
  }) async {
    final response = await _dio.post(
      '/office-hours',
      data: {
        'doctorId': doctorId,
        'workLocationId': workLocationId,
        'dayOfWeek': dayOfWeek,
        'startTime': startTime,
        'endTime': endTime,
        'isGlobal': isGlobal,
      },
    );

    if (response.statusCode == 201 || response.statusCode == 200) {
      if (response.data['success'] == true) {
        return OfficeHour.fromJson(response.data['data']);
      }
    }
    return null;
  }

  Future<bool> deleteOfficeHour(String id) async {
    final response = await _dio.delete('/office-hours/$id');

    if (response.statusCode == 200) {
      return response.data['success'] == true;
    }
    return false;
  }
}

File: source/medicalink-fe-staff-develop/medicalink-fe-staff-develop/src/api/services/office-hour.service.ts
/**
 * Office Hours API Service
 * Handles all API calls related to office hours/schedules
 * API Base: /api/office-hours
 */
import { apiClient } from '../core/client'
import type { PaginationParams } from '../types/common.types'

// ============================================================================
// Types
// ============================================================================

export interface OfficeHour {
  id: string
  doctorId: string | null
  workLocationId: string | null
  dayOfWeek: number // 0-6 (0 = Sunday, 1 = Monday, ..., 6 = Saturday)
  startTime: string // HH:mm format or ISO DateTime
  endTime: string // HH:mm format or ISO DateTime
  isGlobal: boolean
  createdAt: string
  updatedAt: string
  // Expanded relations (optional, populated by backend)
  doctor?: {
    id: string
    firstName: string
    lastName: string
    specialtyName?: string
  }
  workLocation?: {
    id: string
    name: string
  }
}

export interface OfficeHourQueryParams extends PaginationParams {
  doctorId?: string
  workLocationId?: string
}

export interface CreateOfficeHourRequest {
  doctorId?: string | null
  workLocationId?: string | null
  dayOfWeek: number
  startTime: string // HH:mm format
  endTime: string // HH:mm format
  isGlobal?: boolean
}

export interface OfficeHoursGroupedResponse {
  global: OfficeHour[]
  workLocation: OfficeHour[]
  doctor: OfficeHour[]
  doctorInLocation: OfficeHour[]
}

export interface OfficeHoursApiResponse {
  success: boolean
  message: string
  data: OfficeHoursGroupedResponse
  timestamp: string
  path: string
  method: string
  statusCode: number
}

// ============================================================================
// Service Class
// ============================================================================

class OfficeHourService {
  // --------------------------------------------------------------------------
  // Office Hours CRUD
  // --------------------------------------------------------------------------

  /**
   * Get all office hours with filtering (staff only)
   * GET /api/office-hours
   */
  async getOfficeHours(
    params: OfficeHourQueryParams = {}
  ): Promise<OfficeHoursApiResponse> {
    const response = await apiClient.get<OfficeHoursApiResponse>(
      '/office-hours',
      { params }
    )
    return response.data
  }

  /**
   * Create new office hours entry (staff only)
   * POST /api/office-hours
   */
  async createOfficeHour(
    data: CreateOfficeHourRequest
  ): Promise<{ success: boolean; data: OfficeHour }> {
    const response = await apiClient.post<{
      success: boolean
      data: OfficeHour
    }>('/office-hours', data)
    return response.data
  }

  /**
   * Delete an office hours entry (staff only)
   * DELETE /api/office-hours/:id
   */
  async deleteOfficeHour(
    id: string
  ): Promise<{ success: boolean; message: string }> {
    const response = await apiClient.delete<{
      success: boolean
      message: string
    }>(`/office-hours/${id}`)
    return response.data
  }
}

// ============================================================================
// Export singleton instance
// ============================================================================

export const officeHourService = new OfficeHourService()
